webapp/src/main/java/org/apache/atlas/web/rest/TypeCacheRefreshREST.java

```diff
@@ -28,46 +28,50 @@ public class TypeCacheRefreshREST {
     private static final Logger LOG = LoggerFactory.getLogger(TypeCacheRefreshREST.class);
 
     private final AtlasTypeDefStore typeDefStore;
-    private final GraphBackedSearchIndexer graphBackedSearchIndexer;
-    private final AtlasTypeRegistry typeRegistry;
+    private final IAtlasGraphProvider provider;
 
     @Inject
-    public TypeCacheRefreshREST(AtlasTypeDefStore typeDefStore, GraphBackedSearchIndexer graphBackedSearchIndexer, AtlasTypeRegistry typeRegistry) {
+    public TypeCacheRefreshREST(AtlasTypeDefStore typeDefStore, IAtlasGraphProvider provider) {
         this.typeDefStore = typeDefStore;
-        this.graphBackedSearchIndexer = graphBackedSearchIndexer;
-        this.typeRegistry = typeRegistry;
+        this.provider = provider;
     }
 
     /**
      * API to refresh type-def cache.
+     *
      * @throws AtlasBaseException
      * @HTTP 204 if type def cache is refreshed successfully
      * @HTTP 500 if there is an error refreshing type def cache
      */
     @POST
     @Path("/refresh")
     @Timed
-    public void refreshCache() throws AtlasBaseException {
-        LOG.info("Initiating type-def cache refresh");
+    public void refreshCache(@QueryParam("expectedFieldKeys") int expectedFieldKeys) throws AtlasBaseException, RepositoryException, InterruptedException {
+        LOG.info("Initiating type-def cache refresh with expectedFieldKeys = {}", expectedFieldKeys);
+        int currentSize = provider.get().getManagementSystem().getGraphIndex(VERTEX_INDEX).getFieldKeys().size();
+        LOG.info("Size of field keys before refresh = {}", currentSize);
+
+        long totalWaitTimeInMillis = 15 * 1000;//15 seconds
+        long sleepTimeInMillis = 500;
+        long totalIterationsAllowed = Math.floorDiv(totalWaitTimeInMillis, sleepTimeInMillis);
+        int counter = 0;
 
+        while (currentSize != expectedFieldKeys && counter++ < totalIterationsAllowed) {
+            currentSize = provider.get().getManagementSystem().getGraphIndex(VERTEX_INDEX).getFieldKeys().size();
+            LOG.info("Size found = {} at iteration {}", currentSize, counter);
+            Thread.sleep(sleepTimeInMillis);
+        }
+        //This condition will hold true when expected fieldKeys did not appear even after waiting for totalWaitTimeInMillis
+        if (counter > totalIterationsAllowed) {
+            LOG.warn("Could not find desired count of fieldKeys {} after {} ms of wait", expectedFieldKeys, totalWaitTimeInMillis);
```

Comment:
Change to Log.error for now, so we search on Sentry

Reply:
Incorporated change